[WebGPU] RemoteBuffer::copy can result in an out of bounds write

mwyrzykowski opened the pull request:
#### 0fa174e18a820faa0c61c796fabe4288683a81d2
<pre>
[WebGPU] RemoteBuffer::copy can result in an out of bounds write
<a href="https://bugs.webkit.org/show_bug.cgi?id=274678">https://bugs.webkit.org/show_bug.cgi?id=274678</a>
&lt;radar://128591350&gt;

Reviewed by Tadeu Zagallo.

RemoteBuffer::copy should use the actual buffer size, not the initial size.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::getBufferContents):
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferGetActualSize):
* Source/WebGPU/WebGPU/WebGPU.h:

* LayoutTests/fast/webgpu/regression/repro_274678-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_274678.html: Added.
Add regression test for running locally.

Canonical link: <a href="https://commits.webkit.org/279368@main">https://commits.webkit.org/279368@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48d31a64262d89e9c4f54accdc04938c686836f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53319 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5817 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4045 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55625 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40019 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43232 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2645 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55417 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46056 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24362 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3373 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2201 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3540 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58194 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3529 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50632 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29683 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46262 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11613 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30607 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29446 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
```diff
@@ -89,7 +89,7 @@ auto BufferImpl::getBufferContents() -> MappedRange
 
     auto* pointer = wgpuBufferGetBufferContents(m_backing.get());
-    auto bufferSize = wgpuBufferGetSize(m_backing.get());
+    auto bufferSize = wgpuBufferGetActualSize(m_backing.get());
```
Would it be clearer if these were called `wgpuBufferGetInitialSize` and `wgpuBufferGetCurrentSize`?

mwyrzykowski committed:
[WebGPU] RemoteBuffer::copy can result in an out of bounds write
https://bugs.webkit.org/show_bug.cgi?id=274678
<radar://128591350>

Reviewed by Tadeu Zagallo.

RemoteBuffer::copy should use the actual buffer size, not the initial size.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUBufferImpl.cpp:
(WebCore::WebGPU::BufferImpl::getBufferContents):
* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferGetActualSize):
* Source/WebGPU/WebGPU/WebGPU.h:

* LayoutTests/fast/webgpu/regression/repro_274678-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_274678.html: Added.
Add regression test for running locally.

Canonical link: https://commits.webkit.org/279368@main
